src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/UpsertDocument.cs

```diff
@@ -35,135 +35,149 @@ NpgsqlTransaction transaction
         {
-            if (pe.SqlState == PostgresErrorCodes.SerializationFailure)
-            {
-                _logger.LogDebug(
-                    pe,
-                    "Transaction conflict on Aliases table insert - {TraceId}",
-                    upsertRequest.TraceId
-                );
-                return new UpsertResult.UpsertFailureWriteConflict();
-            }
-
-            if (pe.SqlState == PostgresErrorCodes.UniqueViolation)
-            {
-                _logger.LogInformation(
-                    "Failure: alias identity already exists - {TraceId}",
-                    upsertRequest.TraceId
-                );
+            _logger.LogInformation(
+                pe,
+                "Failure: alias identity already exists - {TraceId}",
+                upsertRequest.TraceId
+            );
 
-                return new UpsertResult.UpsertFailureIdentityConflict(
-                    upsertRequest.ResourceInfo.ResourceName.Value,
-                    upsertRequest.DocumentInfo.DocumentIdentity.DocumentIdentityElements.Select(d =>
-                        new KeyValuePair<string, string>(d.IdentityJsonPath.Value.Substring(d.IdentityJsonPath.Value.LastIndexOf('.') + 1), d.IdentityValue)
+            return new UpsertResult.UpsertFailureIdentityConflict(
+                upsertRequest.ResourceInfo.ResourceName.Value,
+                upsertRequest.DocumentInfo.DocumentIdentity.DocumentIdentityElements.Select(
+                    d => new KeyValuePair<string, string>(
+                        d.IdentityJsonPath.Value.Substring(d.IdentityJsonPath.Value.LastIndexOf('.') + 1),
+                        d.IdentityValue
                     )
-                );
-            }
+                )
+            );
+        }
+
+        DocumentReference[] documentReferences = upsertRequest.DocumentInfo.DocumentReferences;
 
-            _logger.LogError(pe, "Failure on Aliases table insert - {TraceId}", upsertRequest.TraceId);
-            return new UpsertResult.UnknownFailure("Upsert failure");
+        if (documentReferences.Length > 0)
+        {
+            // Next insert into References
+            int numberOfRowsInserted = await _sqlAction.InsertReferences(
+                new(
+                    ParentDocumentPartitionKey: documentPartitionKey,
+                    ParentDocumentId: newDocumentId,
+                    ReferentialIds: documentReferences.Select(x => x.ReferentialId.Value).ToArray(),
+                    ReferentialPartitionKeys: documentReferences
+                        .Select(x => PartitionKeyFor(x.ReferentialId).Value)
+                        .ToArray()
+                ),
+                connection,
+                transaction
+            );
+
+            Trace.Assert(
+                numberOfRowsInserted == documentReferences.Length,
+                "Database did not insert all references"
```

Comment:
Does this throw an exception?

Reply:
Yes, this would throw an exception. It would be exceeding weird to be violated. But we get back a rows inserted count from the db, and are using that fancy `unnest` so may as well assert it in case somehow we aren't using it correctly.

Reply:
Yes, exceedingly weird for that to occur, hence I want to be sure that we know how the response actually works. The [Trace.Assert](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.assert?view=net-8.0) documentation says that it will pop up a message box, and doesn't mention throwing an exception.

I think we need to change to just throwing an exception instead of using `Trace.Assert`. I just did a little test and confirmed that no exception occurs.

![image](https://github.com/Ed-Fi-Alliance-OSS/Data-Management-Service/assets/9324390/700d77c1-a118-48c2-807d-c2f0570ed475)

Reply:
@JBrenesSimpat